packages/flutter/test/material/elevated_button_test.dart

```diff
@@ -2400,11 +2406,103 @@ void main() {
     // The icon is aligned to the left of the button.
     expect(buttonTopLeft.dx, iconTopLeft.dx - 24.0); // 24.0 - padding between icon and button edge.
   });
-}
 
-TextStyle _iconStyle(WidgetTester tester, IconData icon) {
-  final RichText iconRichText = tester.widget<RichText>(
-    find.descendant(of: find.byIcon(icon), matching: find.byType(RichText)),
-  );
-  return iconRichText.text.style!;
+  // This is a test regression for https://github.com/flutter/flutter/issues/154798.
+  testWidgets('ElevatedButton.styleFrom can customize the button icon', (WidgetTester tester) async {
+    const Color iconColor = Color(0xFFF000FF);
+    const double iconSize = 32.0;
+    const Color disabledIconColor = Color(0xFFFFF000);
+    Widget buildButton({ bool enabled = true }) {
+      return MaterialApp(
+        home: Material(
+          child: ElevatedButton.icon(
+            style: ElevatedButton.styleFrom(
+              iconColor: iconColor,
+              iconSize: iconSize,
+              disabledIconColor: disabledIconColor,
+            ),
+            onPressed: enabled ? () {} : null,
+            icon: const Icon(Icons.add),
+            label: const Text('Button'),
+          ),
+        ),
+      );
+    }
+
+    // Test enabled button.
+    await tester.pumpWidget(buildButton());
+    expect(tester.getSize(find.byIcon(Icons.add)), const Size(iconSize, iconSize));
+    expect(iconStyle(tester, Icons.add).color, iconColor);
+
+    // Test disabled button.
+    await tester.pumpWidget(buildButton(enabled: false));
+    expect(iconStyle(tester, Icons.add).color, disabledIconColor);
+  });
+
+  // This is a test regression for https://github.com/flutter/flutter/issues/154798.
+  test('ElevatedButton.styleFrom constructor', () {
```

Comment:
Just a question, what is this unit test testing for?

Reply:
This tests all the expected styleFrom method parameters. Some existing tests only some of the styleFrom but not every parameter hence we missed some icon properties.

This test is regular test not a widget tests just to props similar to debug fill properties tests.

Reply:
What if you create an additional ButtonStyle with the same parameters but with the normal ButtonStyle constructor, and then compare their equality? Not sure if that would work like I'm thinking it would, but if so it would fail if an additional parameter is added to ButtonStyle but not to styleFrom, which would be pretty valuable.

Reply:
@justinmc  Something like this?

```dart
  // Regression for https://github.com/flutter/flutter/issues/154798.
  test('ElevatedButton.styleFrom constructor', () {
    const Color foregroundColor = Color(0xFF000022);
    const Color backgroundColor = Color(0xFF000011);
    const Color disabledForegroundColor = Color(0xFF000033);
    const Color disabledBackgroundColor = Color(0xFF000044);
    const Color shadowColor = Color(0xFF000066);
    const Color surfaceTintColor = Color(0xFF000077);
    const Color iconColor = Color(0xFF000088);
    const double iconSize = 32.0;
    const Color disabledIconColor = Color(0xFF000099);
    const Color overlayColor = Color(0xFF000055);
    const double elevation = 10.0;
    const TextStyle textStyle = TextStyle(fontSize: 21.0);
    const EdgeInsetsGeometry padding = EdgeInsets.all(16.0);
    const Size minimumSize = Size(88.0, 36.0);
    const Size fixedSize = Size(100.0, 50.0);
    const Size maximumSize = Size(120.0, 48.0);
    const BorderSide side = BorderSide(width: 2.0, color: Color(0xFF0000AA));
    final OutlinedBorder shape = RoundedRectangleBorder(borderRadius: BorderRadius.circular(8.0));
    const MouseCursor enabledMouseCursor = SystemMouseCursors.click;
    const MouseCursor disabledMouseCursor = SystemMouseCursors.basic;
    const VisualDensity visualDensity = VisualDensity.compact;
    const MaterialTapTargetSize tapTargetSize = MaterialTapTargetSize.shrinkWrap;
    const Duration animationDuration = Duration(milliseconds: 200);
    const bool enableFeedback = true;
    const AlignmentGeometry alignment = Alignment.center;
    const InteractiveInkFeatureFactory splashFactory = InkRipple.splashFactory;
    Widget backgroundBuilder(BuildContext context, Set<WidgetState> states, Widget? child) {
      return const SizedBox();
    }
    Widget foregroundBuilder(BuildContext context, Set<WidgetState> states, Widget? child) {
      return const SizedBox();
    }

    final ButtonStyle styleFrom = ElevatedButton.styleFrom(
      foregroundColor: foregroundColor,
      backgroundColor: backgroundColor,
      disabledForegroundColor: disabledForegroundColor,
      disabledBackgroundColor: disabledBackgroundColor,
      shadowColor: shadowColor,
      surfaceTintColor: surfaceTintColor,
      iconColor: iconColor,
      iconSize: iconSize,
      disabledIconColor: disabledIconColor,
      overlayColor: overlayColor,
      elevation: elevation,
      textStyle: textStyle,
      padding: padding,
      minimumSize: minimumSize,
      fixedSize: fixedSize,
      maximumSize: maximumSize,
      side: side,
      shape: shape,
      enabledMouseCursor: enabledMouseCursor,
      disabledMouseCursor: disabledMouseCursor,
      visualDensity: visualDensity,
      tapTargetSize: tapTargetSize,
      animationDuration: animationDuration,
      enableFeedback: enableFeedback,
      alignment: alignment,
      splashFactory: splashFactory,
      backgroundBuilder: backgroundBuilder,
      foregroundBuilder: foregroundBuilder,
    );

    final ButtonStyle buttonStyle = ButtonStyle(
      foregroundColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledForegroundColor,
          WidgetState.any: foregroundColor,
        },
      ),
      backgroundColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledBackgroundColor,
          WidgetState.any: backgroundColor,
        },
      ),
      shadowColor: const WidgetStatePropertyAll<Color?>(shadowColor),
      surfaceTintColor: const WidgetStatePropertyAll<Color?>(surfaceTintColor),
      iconColor: const WidgetStateProperty<Color?>.fromMap(
        <WidgetStatesConstraint, Color?>{
          WidgetState.disabled: disabledIconColor,
          WidgetState.any: iconColor,
        },
      ),
      iconSize: const WidgetStatePropertyAll<double>(iconSize),
      overlayColor: const WidgetStatePropertyAll<Color?>(overlayColor),
      elevation: const WidgetStatePropertyAll<double>(elevation),
      textStyle: const WidgetStatePropertyAll<TextStyle?>(textStyle),
      padding: const WidgetStatePropertyAll<EdgeInsetsGeometry>(padding),
      minimumSize: const WidgetStatePropertyAll<Size>(minimumSize),
      fixedSize: const WidgetStatePropertyAll<Size>(fixedSize),
      maximumSize: const WidgetStatePropertyAll<Size>(maximumSize),
      side: const WidgetStatePropertyAll<BorderSide>(side),
      shape: WidgetStatePropertyAll<OutlinedBorder>(shape),
      mouseCursor: const WidgetStateProperty<MouseCursor?>.fromMap(
        <WidgetStatesConstraint, MouseCursor?>{
          WidgetState.disabled: disabledMouseCursor,
          WidgetState.any: enabledMouseCursor,
        },
      ),
      visualDensity: visualDensity,
      tapTargetSize: tapTargetSize,
      animationDuration: animationDuration,
      enableFeedback: enableFeedback,
      alignment: alignment,
      splashFactory: splashFactory,
      backgroundBuilder: backgroundBuilder,
      foregroundBuilder: foregroundBuilder,
    );

    expect(
      styleFrom.foregroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.foregroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.foregroundColor!.resolve(<WidgetState>{}),
      buttonStyle.foregroundColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.backgroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.backgroundColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.backgroundColor!.resolve(<WidgetState>{}),
      buttonStyle.backgroundColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.shadowColor!.resolve(<WidgetState>{}),
      buttonStyle.shadowColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.surfaceTintColor!.resolve(<WidgetState>{}),
      buttonStyle.surfaceTintColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.iconColor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.iconColor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.iconColor!.resolve(<WidgetState>{}),
      buttonStyle.iconColor!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.iconSize!.resolve(<WidgetState>{}),
      buttonStyle.iconSize!.resolve(<WidgetState>{}),
    );
    expect(styleFrom.overlayColor!.resolve(<WidgetState>{}), null); // This is null when a foreground is provided.
    expect(
      styleFrom.elevation!.resolve(<WidgetState>{}),
      buttonStyle.elevation!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.textStyle!.resolve(<WidgetState>{}),
      buttonStyle.textStyle!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.padding!.resolve(<WidgetState>{}),
      buttonStyle.padding!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.minimumSize!.resolve(<WidgetState>{}),
      buttonStyle.minimumSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.fixedSize!.resolve(<WidgetState>{}),
      buttonStyle.fixedSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.maximumSize!.resolve(<WidgetState>{}),
      buttonStyle.maximumSize!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.side!.resolve(<WidgetState>{}),
      buttonStyle.side!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.shape!.resolve(<WidgetState>{}),
      buttonStyle.shape!.resolve(<WidgetState>{}),
    );
    expect(
      styleFrom.mouseCursor!.resolve(<WidgetState>{ WidgetState.disabled }),
      buttonStyle.mouseCursor!.resolve(<WidgetState>{ WidgetState.disabled }),
    );
    expect(
      styleFrom.mouseCursor!.resolve(<WidgetState>{}),
      buttonStyle.mouseCursor!.resolve(<WidgetState>{}),
    );
    expect(styleFrom.visualDensity, buttonStyle.visualDensity);
    expect(styleFrom.tapTargetSize, buttonStyle.tapTargetSize);
    expect(styleFrom.animationDuration, buttonStyle.animationDuration);
    expect(styleFrom.enableFeedback, buttonStyle.enableFeedback);
    expect(styleFrom.alignment, buttonStyle.alignment);
    expect(styleFrom.splashFactory, buttonStyle.splashFactory);
    expect(styleFrom.backgroundBuilder, buttonStyle.backgroundBuilder);
    expect(styleFrom.foregroundBuilder, buttonStyle.foregroundBuilder);
  });
```

Reply:
My guess is that it would just need 1 equality check:

```dart
expect(styleFrom == buttonStyle, isTrue);
```

Though if we were to add a `ButtonStyle` property without updating this test, they would both be initialized to `null` and we likely wouldn't be able to catch that `.styleFrom()` was missing something.

Reply:
`expect(styleFrom == buttonStyle, isTrue);` is same as `expect(styleFrom, buttonStyle);`. 

I tried this already. Either checks won't work  as the some of the individuals properties return different types. We will have to break down each properties like i did in the https://github.com/flutter/flutter/pull/154821#discussion_r1751495167.